### ISSUE TITLE
Workflow & Deployment for Talk

### ISSUE THREAD

**Issue by campaignupgrade** [coralproject/talk#2074]
#### Do you want to request a **feature** or report a **bug**?

Feature — documentation enhancement? 

#### Intended outcome:

Set up a functional instance of Talk with a deployment and backup. 

I was directed to post here by support@coralproject.zendesk.com. 

I’m the IT lead for Truthout.org, a progressive news site with about two millions pageviews per month. 

We’ll looking to integrate Talk commenting into our WordPress system and we’re currently stalled. We’re not entirely clear on the suggested workflow for modifying and updating Talk and its dependencies over time. 

We have Talk set up on a Puppet-controlled Docker Swarm cluster on Digital Ocean using the supplied Docker method at https://docs.coralproject.net/talk/installation-from-docker/. 

I’m able to successfully login as admin to the moderation area. 

Where we are stalled is in modifying the installation and adding plugins. The documentation at https://docs.coralproject.net/talk/plugins/ is a bit unclear on this, but it appears that we’d want to git checkout each plugin to its own folder in a plugins directory alongside our Dockerfile, if I am reading the `-onbuild` method correctly. 

We also have question regarding database backups as it appears the mongo volume isn’t mounted. 

Is Coral Project available for support on this? We’d really love some advice on the install workflow. 

In short — how exactly are you self-hosting this and what's your workflow for updating and backups?

**Comment by kgardnr** [comment]
Hey @slamorte!

So if you're using custom plugins, definitely go the `onbuild` route. You can manage your `plugins.json`, `plugins` dir, and `Dockerfile` in a git repo, as per the docs:

```This means that you can create a repository for your organization that simply includes the above Dockerfile, a directory of plugins, and a plugins.json file which specifies the activated plugins, and you can deploy Talk easily to your containerized infrastructure. ```

For Mongo, we highly recommend using MongoDB Atlas - we've used it extensively and it really works well to generate and manage a db cluster for Talk. 

Make sure you also leverage the Pre-Launch Checklist - https://docs.coralproject.net/talk/pre-launch-checklist/ - just to make sure you've covering all of the bases.

**Comment by campaignupgrade** [comment]
Thanks for this!

I've already used to the pre-flight checklist to jump start conversation with the editorial team. It's very comprehensive, thanks! Great resource. 

For code, do you recommend CI for deployment? How are you deploying your docker repo to a production server?

For the `plugins` dir, how are you managing plugin versioning? I see each plugin is usually its own repo, and I am curious if you are managing with composer, git submoduling, or some other progress like simple zip checkouts?

**Comment by kgardnr** [comment]
Anytime! 

Yes, we do recommend CI - we use CircleCI, but others use Jenkins, Travis, and others. We're doing manual deploys to our own environments (which are mostly for testing) - at one point though we did have a Slack bot to deploy from master to staging. Clients typically choose manual or automatic deployments depending on their existing deployment strategies and environments. We do recommend having a staging and a prod separately.

For plugin versioning, we've typically recommended git submodules and seen that work really well for clients. If you go another route, let us know, we'd love to add that to our docs for sure.

Let us know if you have more questions!

**Comment by campaignupgrade** [comment]
Still trying to get the -onboard process to work. 

I've created a Dockerfile that contains only the line "FROM coralproject/talk:4.5-onbuild", and I'm building it using the stock coralproject/talk image like this:

`docker build -t coralproject/talk:4.5 --build-arg TALK_DEFAULT_LANG=es .`

There's a few errors thrown, nothing major, but the end result does not produce any running containers. 

```
$ docker container ls
CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
$
```

I do see it results in a new image... do I now need to make a new `docker-compose.yml` that uses this built image during a `docker-compose up -d`? 

I see I'm not the only docker-noob to struggle with this and once I've got this figured out I'd be happy to write clearer steps for the -onboard docs.

**Comment by oafx** [comment]
*onboard = onload 😉 

You're right, docker build only creates an image. I suggest you to tag it with another name like `-t myawesometalkimage:4.5`, so docker won't have any problems with resolving the name. After executing `docker build` you can, for example: 

* run your image with docker run, i.e.: `docker run -d -p 3000:3000 myawesometalkimage:4.5 -e "NODE_ENV=development" -e TALK_MONGO_URL=<your mongo url> [...your other talk environment variables]` _note: the `-d` (detached) flag makes the container run in the background_
* adjust the `docker-compose.yml` from https://docs.coralproject.net/talk/ or create a new docker-compose file, that uses your image: 
```yml
# For details on the syntax of docker-compose.yml files, check out:
# https://docs.docker.com/compose/compose-file/compose-file-v2/

version: '2'
services:
  talk:
    image: myawesometalkimage:4.5
    restart: always
    ports:
      - "3000:3000"

# [...] and so on, see the installation example for a docker-compose.yml in the talk docs
# sidenote: you can also adjust the docker-compose file to build the image on the fly-ish
# https://docs.docker.com/compose/compose-file/compose-file-v2/#build
```
and afterwards run it with `docker-compose up -d` as described in the docs

* or push the image to a container registry (which is AFAIK the way to go, if you want to deploy it remotely)

I reccomend to read some docker intros/tutorials 

Good Luck!
Christian

**Comment by campaignupgrade** [comment]
Thanks Christian!

Yes, I'm pretty new to docker. I've tried reading the docs and man pages but it can be hard to get the basic docker concepts until I experiment with something like these Talk builds. But I think I understand images now. 

We hacked about over the last few days and mostly figured this out for ourselves. We took the path of adjusting the `docker-compose.yml` to use the custom build. Still running into a few issues. 

1.  Does the `plugins.json` file need to specific the default plugins as well, or only my selected optional plugins? 

2. Do coralproject-hosted plugins need to be placed in the /plugins/ directory, or only custom plugins? 

Lastly, once I try run the custom image is hangs at "restarting:"

```
$  docker-compose ps
    Name                  Command                 State        Ports
----------------------------------------------------------------------
talk_mongo_1   docker-entrypoint.sh mongod      Up           27017/tcp
talk_redis_1   docker-entrypoint.sh redis ...   Up           6379/tcp
talk_talk_1    yarn start                       Restarting
```
Any suggestion on troubleshooting? I don't see anything related in the github issues. 

EDIT: Never mind about the restart, it appears related to my plugins.json. I'll add one plugin at a time to troubleshoot. 

EDIT 2: i've answered the first two as well. Yes, and no. =)
